### PR TITLE
Update ScubaGear code for AAD baseline policy 7.4

### PR DESCRIPTION
**Pull request by Sloane4**
#  #

## 🗣 Description ##

- Remove the checks for the PIM settings "Allow permanent active assignment" and "Expire active assignments after 15 days".
- Replace those checks with code that examines if there are any permanent Active assignments. If there are, we show a Fail unless those assignments are in a config file exclusion.
- Also show a list of the roles currently "permanently" assigned to privileged roles. We currently hide this information which makes it harder for admins because they have to go look for it.

## 💭 Motivation and context ##

Code updated to match policy update in the baseline document
Closing #257 

## 🧪 Testing ##

- Unit tests
- Running against G5 test tenant

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Create a release.

### CODE REVIEW

File: Rego/AADConfig.rego [hunk 1]
```diff
@@ -715,37 +715,40 @@ tests[{
...
 #--
-RolesWithoutLimitedExpirationPeriod[Role.DisplayName] {
-    Role := input.privileged_roles[_]
-    Rule := Role.Rules[_]
-    RuleMatch := Rule.Id == "Expiration_Admin_Assignment"
-    ExpirationNotRequired := Rule.AdditionalProperties.isExpirationRequired == false
-    MaximumDurationCorrect := Rule.AdditionalProperties.maximumDuration == "P15D"
+default PrivilegedRolesExclusions(_, _) := false
+PrivilegedRolesExclusions(PrivilegedRoles, PolicyID) := true if {
+    PrivilegedRolesUsers := { x.PrincipalId | some x in PrivilegedRoles.Assignments; x.ActivatedUsing.EndDateTime == null }
```
[the remark this answers is not in the document]
Comment (Sloane4) [reply Rego/AADConfig.rego]: update

File: Rego/AADConfig.rego [hunk 2]
```diff
@@ -715,37 +715,40 @@ tests[{
...
-    Case1 := count([Condition | Condition = Conditions1[_]; Condition == false]) == 0
+    AllowedPrivilegedRolesUsers := { y | some y in input.scuba_config.Aad[PolicyID].RoleExclusions.Users; y != null }
+    AllowedPrivilegedRolesGroups := { y | some y in input.scuba_config.Aad[PolicyID].RoleExclusions.Groups; y != null }
+    AllowedPrivilegedRoles := AllowedPrivilegedRolesUsers | AllowedPrivilegedRolesGroups
 
-    # Role policy requires assignment expiration, but maximum duration is not 15 days
-    Conditions2 := [RuleMatch == true, ExpirationNotRequired == false, MaximumDurationCorrect == false]
-    Case2 := count([Condition | Condition = Conditions2[_]; Condition == false]) == 0
+    count(PrivilegedRolesUsers) > 0
+    count(PrivilegedRolesUsers - AllowedPrivilegedRoles) != 0
```
[the remark this answers is not in the document]
Comment (Sloane4) [reply Rego/AADConfig.rego]: updated

File: Rego/AADConfig.rego [hunk 3]
```diff
@@ -715,37 +715,40 @@ tests[{
...
+}
 
-    # Filter: only include rules that meet one of the two cases
-    Conditions := [Case1, Case2]
-    count([Condition | Condition = Conditions[_]; Condition == true]) > 0
+PrivilegedRolesExclusions(PrivilegedRoles, PolicyID) := true if {
+    count({ x.PrincipalId | some x in PrivilegedRoles.Assignments; x.ActivatedUsing.EndDateTime == null }) > 0
```
[the remark this answers is not in the document]
Comment (Sloane4) [reply Rego/AADConfig.rego]: updated

File: Rego/AADConfig.rego [hunk 4]
```diff
@@ -715,37 +715,40 @@ tests[{
...
+default PrivilegedRolesExclusions(_, _) := false
+PrivilegedRolesExclusions(PrivilegedRoles, PolicyID) := true if {
+    PrivilegedRolesUsers := { x.PrincipalId | some x in PrivilegedRoles.Assignments; x.ActivatedUsing.EndDateTime == null }
 
-    # Role policy does not require assignment expiration
-    Conditions1 := [RuleMatch == true, ExpirationNotRequired == true]
-    Case1 := count([Condition | Condition = Conditions1[_]; Condition == false]) == 0
+    AllowedPrivilegedRolesUsers := { y | some y in input.scuba_config.Aad[PolicyID].RoleExclusions.Users; y != null }
+    AllowedPrivilegedRolesGroups := { y | some y in input.scuba_config.Aad[PolicyID].RoleExclusions.Groups; y != null }
+    AllowedPrivilegedRoles := AllowedPrivilegedRolesUsers | AllowedPrivilegedRolesGroups
```
[the remark this answers is not in the document]
Comment (Sloane4) [reply Rego/AADConfig.rego]: updated

File: Rego/AADConfig.rego [hunk 5]
```diff
@@ -735,37 +735,40 @@ tests[{
...
 }
 
 tests[{
     "PolicyId" : "MS.AAD.7.4v1",
     "Criticality" : "Shall",
     "Commandlet" : ["Get-MgSubscribedSku", "Get-PrivilegedRole"],
-    "ActualValue" : RolesWithoutLimitedExpirationPeriod,
-    "ReportDetails" : ReportDetailsArrayLicenseWarning(RolesWithoutLimitedExpirationPeriod, DescriptionString),
+    "ActualValue" : PrivilegedRolesWithoutExpirationPeriod,
+    "ReportDetails" : ReportDetailsArrayLicenseWarning(PrivilegedRolesWithoutExpirationPeriod, DescriptionString),
```
Comment (crutchfield) [inline Rego/AADConfig.rego:767]: Function name, ReportDetailsArrayLiscenseWarning,  does not appear to be a real meaningful contribution to readability.

**Comment by tkol2022** [comment]
### Test Results
I tested the code and reviewed the Rego code. It appears to be working right based on the report output however I found a discrepancy in the Rego code. Let me know if this makes sense and if I missed something. Thanks.

The Rego code for 7.4 is currently looking at the x.ActivatedUsing.EndDateTime field, but it should be looking at the x.EndDateTime field (similar to how we reference  x.StartDateTime in the code for policy 7.5). When I created a test active assignment in the E5 tenant that has an expiration date and reviewed the provider JSON, I saw that the ActivatedUsing field does NOT contain the expiration date (instead the EndDateTime at the root of the JSON structure node is used).

image
image

image

**Comment by crutchfield** [comment]
@Sloane4 Questions: for MS.AAD7.4v1:

1) Does this phrase, "except for emergency and service accounts",  indicate the  account should be listed as excluded?

2) Is it acceptable to have 'alternative PAM system' instead of P2 license?

**Review by crutchfield — APPROVED** [review]

**Comment by Sloane4** [comment]
@nanda-katikaneni ready for merge

**Pull request merged**